virtool/tasks/runner.py

```diff
@@ -14,29 +13,48 @@
 
 class TaskRunner:
     def __init__(
-        self, data: DataLayer, tasks_client: AbstractTasksClient, app: Application
+        self, tasks_client: AbstractTasksClient, app: Application
     ):
-        self._data = data
         self._tasks_client = tasks_client
         self.app = app
+        self._current_task: Task = None
 
     async def run(self):
-        logging.info("Started task runner")
-
         try:
             while True:
                 logging.info("Waiting for next task")
-
                 task_id = await self._tasks_client.pop()
 
                 await self.run_task(task_id)
 
+                await asyncio.shield(self._current_task)
+
+                self._current_task = None
+
                 logging.info("Finished task: %s", task_id)
 
         except asyncio.CancelledError:
-            logging.info("Stopped task runner")
+            if self._current_task:
+                logging.info("Recieved stop signal")
+
+                try:
+                    logging.info("Waiting for task to finish: %s", task_id)
+
+                    await asyncio.wait_for(asyncio.shield(self._current_task), 600)
```

Comment:
This shield concerns me. My impression is that we should only see one cancellation error. Ideally at this point you have already caught any cancelation errors that are being propagated. If your still seeing cancellation errors at this point then we need to think this through, as an error raised at this point would immediately bail from the runner but allow the task to keep running. If that happens good chance aiohttp will shutdown the database connections before the task is done.